### PR TITLE
[IT567] Automatic value for the inventory field.

### PR DESCRIPTION
**Pull request by arek1-618**
Automatic value for the inventory field.

### CODE REVIEW

File: modules/Vtiger/inventoryfields/Basic.php [hunk 1]
```diff
@@ -430,6 +456,20 @@ public function validate($value, string $columnName, bool $isUserFormat)
...
 		}
+		$this->validateIfAutomaticValue();
+	}
+
+	/**
+	 * Validate if the automatic value.
+	 *
+	 * @return void
+	 */
+	protected function validateIfAutomaticValue()
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: remove

File: modules/Vtiger/inventoryfields/Basic.php [hunk 2]
```diff
@@ -466,10 +506,11 @@ public function isRequired()
 	 * @throws \App\Exceptions\AppException
 	 * @throws \App\Exceptions\Security
 	 */
-	public function setValueToRecord(\Vtiger_Record_Model $recordModel, array $item, bool $userFormat)
+	public function setValueToRecord(Vtiger_Record_Model $recordModel, array $item, bool $userFormat)
 	{
 		$column = $this->getColumnName();
-		$value = $item[$column];
+		$this->item = $item;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: ?

File: app/Inventory.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Arkadiusz Adach <a.adach@yetiforce.com>
+ */
+
+namespace App;
+
+/**
+ * Inventory class. Calculate net, gross, margins.
+ */
+class Inventory
```
Comment (rskrzypczak) [inline app/Inventory.php]: ?

File: modules/Vtiger/inventoryfields/Discount.php [hunk 4]
```diff
@@ -69,8 +74,32 @@ public function validate($value, string $columnName, bool $isUserFormat)
...
+			}
 		} elseif (App\TextParser::getTextLength($value) > $this->customMaximumLength[$columnName]) {
 			throw new \App\Exceptions\Security("ERR_VALUE_IS_TOO_LONG||$columnName||$value", 406);
 		}
 	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getAutomaticValue(array $item, bool $userFormat = false)
```
Comment (TomaszQr) [inline modules/Vtiger/inventoryfields/Discount.php]: ?

File: modules/Vtiger/inventoryfields/Basic.php [hunk 5]
```diff
@@ -78,6 +85,51 @@ public static function getInstance(string $moduleName, ?string $type = 'Basic')
...
+	}
+
+	/**
+	 * Round method.
+	 *
+	 * @param float $value
+	 *
+	 * @return float
+	 */
+	public function roundMethod(float $value): float
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/inventoryfields/Basic.php]: unnecessary function

File: modules/Vtiger/inventoryfields/Basic.php [hunk 6]
```diff
@@ -88,6 +140,16 @@ public function getModuleName(): string
 		return $this->moduleName;
 	}
 
+	/**
+	 * Is set automatically.
+	 *
+	 * @return bool
+	 */
+	public function getIsAutomaticValue(): bool
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/inventoryfields/Basic.php]: unused function

File: modules/Vtiger/inventoryfields/Basic.php [hunk 7]
```diff
@@ -126,6 +188,18 @@ public function getDBType()
...
 	}
 
+	/**
+	 * Get the automatic value.
+	 *
+	 * @param array $item
+	 *
+	 * @return mixed
+	 */
+	public function getAutomaticValue(array $item, bool $userFormat = false)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/inventoryfields/Basic.php]:
```suggestion
	public function getValueForSave(array $item, bool $userFormat = false)
```

File: modules/Vtiger/inventoryfields/Basic.php (hunk 5 above)
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: remove

File: modules/Vtiger/inventoryfields/Tax.php [hunk 8]
```diff
@@ -71,7 +76,7 @@ public function getDBValue($value, ?string $name = '')
 	/**
 	 * {@inheritdoc}
 	 */
-	public function validate($value, string $columnName, bool $isUserFormat)
+	protected function validate($value, string $columnName, bool $isUserFormat, $originalValue)
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Tax.php]: revert 'public'

File: modules/Vtiger/inventoryfields/Basic.php (hunk 6 above)
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: remove

File: modules/Vtiger/inventoryfields/Discount.php [hunk 9]
```diff
@@ -30,6 +30,11 @@ class Vtiger_Discount_InventoryField extends Vtiger_Basic_InventoryField
 		'discountparam' => App\Purifier::TEXT
 	];
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Discount.php]: remove

File: modules/Vtiger/inventoryfields/GrossPrice.php [hunk 10]
```diff
@@ -21,6 +21,11 @@ class Vtiger_GrossPrice_InventoryField extends Vtiger_Basic_InventoryField
 	protected $maximumLength = '99999999999999999999';
 	protected $purifyType = \App\Purifier::NUMBER;
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/GrossPrice.php]: remove

File: modules/Vtiger/inventoryfields/Margin.php [hunk 11]
```diff
@@ -21,6 +21,11 @@ class Vtiger_Margin_InventoryField extends Vtiger_Basic_InventoryField
 	protected $maximumLength = '99999999999999999999';
 	protected $purifyType = \App\Purifier::NUMBER;
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Margin.php]: remove

File: modules/Vtiger/inventoryfields/MarginP.php [hunk 12]
```diff
@@ -22,6 +22,11 @@ class Vtiger_MarginP_InventoryField extends Vtiger_Basic_InventoryField
 	protected $maximumLength = '99999999999999999999';
 	protected $purifyType = \App\Purifier::NUMBER;
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/MarginP.php]: remove

File: modules/Vtiger/inventoryfields/NetPrice.php [hunk 13]
```diff
@@ -21,6 +21,11 @@ class Vtiger_NetPrice_InventoryField extends Vtiger_Basic_InventoryField
 	protected $maximumLength = '99999999999999999999';
 	protected $purifyType = \App\Purifier::NUMBER;
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/NetPrice.php]: remove

File: modules/Vtiger/inventoryfields/Tax.php [hunk 14]
```diff
@@ -34,6 +34,11 @@ class Vtiger_Tax_InventoryField extends Vtiger_Basic_InventoryField
 	 */
 	public $shared = ['taxparam' => 'tax_percent'];
 
+	/**
+	 * {@inheritdoc}
+	 */
+	protected $isAutomaticValue = true;
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Tax.php]: remove itd.

File: modules/Vtiger/inventoryfields/Basic.php (hunk 7 above)
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: ?
remove

File: modules/Vtiger/inventoryfields/Basic.php [hunk 15]
```diff
@@ -78,6 +85,51 @@ public static function getInstance(string $moduleName, ?string $type = 'Basic')
...
+	/**
+	 * Calculate from field.
+	 *
+	 * @param string $moduleName
+	 * @param string $type
+	 * @param array  $item
+	 *
+	 * @return float
+	 */
+	public static function calculateFromField(string $moduleName, string $type, array $item, bool $userFormat = false): float
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/inventoryfields/Basic.php]:
```suggestion
	protected function valueFromField(string $type, array $item, bool $userFormat = false): float
```

File: modules/Vtiger/inventoryfields/Basic.php [hunk 16]
```diff
@@ -78,6 +85,51 @@ public static function getInstance(string $moduleName, ?string $type = 'Basic')
...
+	{
+		return self::getInstance($moduleName, $type)->getAutomaticValue($item, $userFormat);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/inventoryfields/Basic.php]:
```suggestion
		return self::getInstance($this->getModuleName(), $type)->getAutomaticValue($item, $userFormat);
```

File: modules/Vtiger/inventoryfields/Basic.php (hunk 15 above)
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]: remove

File: modules/Vtiger/inventoryfields/Basic.php [hunk 17]
```diff
@@ -126,6 +169,18 @@ public function getDBType()
...
+	/**
+	 * Get the automatic value for save.
+	 *
+	 * @param array $item
+	 *
+	 * @return mixed
+	 */
+	public function getValueForSave(array $item, bool $userFormat = false)
+	{
+		return $item[$this->getColumnName()] ?? $this->getDefaultValue();
```
Comment (rskrzypczak) [inline modules/Vtiger/inventoryfields/Basic.php]:
```suggestion
		return $item[$this->getColumnName()] ?? null;
```

File: public_html/layouts/resources/Fields.js [hunk 18]
```diff
@@ -1481,16 +1481,29 @@ window.App.Fields = {
 			if (!value) {
 				value = 0;
 			}
+			let strDecimal = value.toString().split(".")[1];
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: "."  > '.'

File: public_html/layouts/resources/Fields.js [hunk 19]
```diff
@@ -1481,16 +1481,29 @@ window.App.Fields = {
...
+			let numberOfZerosAtTheEnd = 0;
+			if (typeof strDecimal !== "undefined") {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: "undefined"  > 'undefined'

File: public_html/layouts/resources/Fields.js [hunk 20]
```diff
@@ -1481,16 +1481,29 @@ window.App.Fields = {
...
+				for (let i = strDecimal.length - 1; i > 0; --i) {
+					if (strDecimal[i] == "0") {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: "0"

File: public_html/layouts/resources/Fields.js [hunk 21]
```diff
@@ -1481,16 +1481,29 @@ window.App.Fields = {
...
+						break;
+					}
+				}
+			}
 			value = parseFloat(value);
 			if (fixed) {
 				value = value.toFixed(numberOfDecimal);
 			}
-			let splittedFloat = value.toString().split('.');
+			let splittedFloat = value.toString().split(".");
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: "."

File: public_html/layouts/resources/Fields.js [hunk 22]
```diff
@@ -1481,16 +1481,29 @@ window.App.Fields = {
...
 			let integer = App.Fields.Integer.formatToDisplay(splittedFloat[0]);
 			let decimal = splittedFloat[1];
 			if (numberOfDecimal) {
-				if (CONFIG.truncateTrailingZeros && 0===parseInt(decimal)) {
-					return integer;
+				if (!CONFIG.truncateTrailingZeros && decimal) {
+					for (let i = 0; i < numberOfZerosAtTheEnd && decimal.length < numberOfDecimal; ++i) {
+						decimal += "0";
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: "."

**Pull request merged**